src/System.IO.Pipelines/PipeAwaitable.cs

```diff
@@ -10,16 +10,34 @@ internal struct PipeAwaitable
     {
         private static readonly Action _awaitableIsCompleted = () => { };
         private static readonly Action _awaitableIsNotCompleted = () => { };
+        private static readonly Action<object> _cancelOperation = (state) => { };
 
         private CancelledState _cancelledState;
         private Action _state;
+        private CancellationToken _cancellationToken;
+        private CancellationTokenRegistration _cancellationTokenRegistration;
 
         public PipeAwaitable(bool completed)
         {
             _cancelledState = CancelledState.NotCancelled;
             _state = completed ? _awaitableIsCompleted : _awaitableIsNotCompleted;
         }
 
+
+        public void AttachToken(CancellationToken cancellationToken, Action<object> callback, object state)
+        {
+            if (cancellationToken != _cancellationToken)
+            {
+                _cancellationTokenRegistration.Dispose();
```

Comment:
This will deadlock if:

- Dispose is called while the token is being canceled
- AND the registered callback calls a Pipe method the attempts to acquire the _sync lock.

This seems like a realistic enough scenario to warrant using [SafeRegister](https://github.com/aspnet/KestrelHttpServer/blob/rel/1.1.1/src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/CancellationTokenExtensions.cs#L11).